### PR TITLE
[confighttp] newDefault for confighttp package structs

### LINKED ISSUE #9655: [confighttp] Add `NewDefault*` functions for all config structs

### PR DESCRIPTION
**Pull request by Sanket-0510**
**Description:**  
NewDefault methods to all the exported structs in confighttp package

**Link to tracking Issue:** 
closes #9655 

**Testing:** 
Added test for the same.

**Documentation:**
 godoc

### CODE REVIEW

**Comment by Sanket-0510** [comment]
why are contrib-tests failing? : |

File: .chloggen/confighttp-add-newdefault-func.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
```
Comment (astencel-sumo) [inline .chloggen/confighttp-add-newdefault-func.yaml:4]: Why is this a breaking change?

Comment (Sanket-0510) [reply .chloggen/confighttp-add-newdefault-func.yaml:4]: wait it should be enhancement

File: config/confighttp/confighttp.go [hunk 2]
```diff
@@ -276,6 +276,11 @@ type ServerConfig struct {
 	ResponseHeaders map[string]configopaque.String `mapstructure:"response_headers"`
 }
 
+// NewDefaultServerConfig creates new ServerConfig with default values set
+func NewDefaultServerConfig() ServerConfig {
+	return ServerConfig{}
```
[the remark this answers is not in the document]
Comment (Sanket-0510) [reply config/confighttp/confighttp.go:293]: I have added the check for ClientConfig also, I hope this fixes the panic

File: config/confighttp/confighttp.go [hunk 3]
```diff
@@ -105,10 +105,15 @@ type ClientConfig struct {
...
 	// The default values are taken from the values of 'DefaultTransport' of 'http' package.
+
+	// Configure with the non-nil map value.
+	headers := make(map[string]configopaque.String)
+
 	maxIdleConns := 100
 	idleConnTimeout := 90 * time.Second
 
 	return ClientConfig{
+		Headers:         headers,
```
Comment (astencel-sumo) [inline config/confighttp/confighttp.go:116]: Why instantiate a variable `headers`?
```suggestion
		Headers:         make(map[string]configopaque.String),
```

File: config/confighttp/confighttp.go [hunk 4]
```diff
@@ -276,6 +281,16 @@ type ServerConfig struct {
 	ResponseHeaders map[string]configopaque.String `mapstructure:"response_headers"`
 }
 
+// NewDefaultServerConfig creates new ServerConfig with default values set
+func NewDefaultServerConfig() ServerConfig {
+	// Create a new map for ResponseHeaders
+	responseHeaders := make(map[string]configopaque.String)
+
+	return ServerConfig{
+		ResponseHeaders: responseHeaders,
```
Comment (astencel-sumo) [inline config/confighttp/confighttp.go:290]:
```suggestion
		ResponseHeaders: make(map[string]configopaque.String),
```

File: config/confighttp/confighttp.go (hunk 3 above)
Comment (Sanket-0510) [reply config/confighttp/confighttp.go]: Yes it will be just waste of space and memory. P.S - corrected it :sweat_smile:

**Review by astencel-sumo — APPROVED** [review]

**Comment by Sanket-0510** [comment]
@astencel-sumo some contrib tests are failing in github actions, but I did ran all tests on local but nothing failed there, I also looked into the failing test it is expecting nil value for our map.

**Review by TylerHelmuth — COMMENTED** [review]
`ClientConfig` and `ServerConfig` depend on `configtls`.  I think we should wait on this PR until the `NewDefault` functions exist for `configtls`.

**Comment by astencel-sumo** [comment]
> `ClientConfig` and `ServerConfig` depend on `configtls`. I think we should wait on this PR until the `NewDefault` functions exist for `configtls`.

I don't think that's the problem. The reason the contrib tests are failing is that I asked Sanket to make the `NewDefaultClientConfig` and `NewDefaultServerConfig` to initialize the `Headers` and `ResponseHeaders` maps to empty maps. Previously those where `nil`. The tests in contrib expect those properties to be `nil`.

See e.g. here: https://github.com/open-telemetry/opentelemetry-collector/actions/runs/8231961010/job/22508354464?pr=9672#step:4:1109

```console
=== FAIL: . TestLoadConfig/zipkin/2 (0.00s)
    config_test.go:86: 
        	Error Trace:	/tmp/opentelemetry-collector-contrib/exporter/zipkinexporter/config_test.go:86
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -39,3 +39,4 @@
        	            	   Timeout: (time.Duration) 5000000000,
        	            	-  Headers: (map[string]configopaque.String) <nil>,
        	            	+  Headers: (map[string]configopaque.String) {
        	            	+  },
```
I'm not exactly sure what's the preferred course of action here.

- Allow contrib tests to fail and fix them after updating core in contrib?
- Fix those tests before merging this change so that those tests to not check that `Response` and `ResponseHeaders` are `nil`?
- ...?

What's your opinion @TylerHelmuth ?

**Comment by TylerHelmuth** [comment]
Oh my comment was only a general statement, not a reason for the failing tests.  I dont think we should introduce `NewDefault*` functions here until they can depend on the `NewDefault*` functions from `configtls`.

For the tests themselves I'd say we have 2 options:
1. Not use empty map as the default value and instead use nil
2. Allow contrib tests to fail and fix them after updating core in contrib

**Comment by astencel-sumo** [comment]
> Oh my comment was only a general statement, not a reason for the failing tests. I dont think we should introduce `NewDefault*` functions here until they can depend on the `NewDefault*` functions from `configtls`.

Oh, I see :smile: Makes sense, @Sanket-0510 this means this PR should wait for https://github.com/open-telemetry/opentelemetry-collector/pull/9658. After https://github.com/open-telemetry/opentelemetry-collector/pull/9658 is merged, you can incorporate changes from `main` (by merging or rebasing) and then we can continue with this.

> For the tests themselves I'd say we have 2 options:
> 
>     1. Not use empty map as the default value and instead use nil
> 
>     2. Allow contrib tests to fail and fix them after updating core in contrib

I don't like 1, I believe the functions should not leave the map as `nil` and expose users to potential `panic`. Let's go with 2 if possible.

**Comment by TylerHelmuth** [comment]
@Sanket-0510 this is ready to use the `configtls. NewDefault*` funcs

File: config/confighttp/confighttp.go [hunk 5]
```diff
@@ -107,10 +107,12 @@ type ClientConfig struct {
 // We encourage to use this function to create an object of ClientConfig.
 func NewDefaultClientConfig() ClientConfig {
 	// The default values are taken from the values of 'DefaultTransport' of 'http' package.
+
```
Comment (TylerHelmuth) [inline config/confighttp/confighttp.go:110]:
```suggestion
```

File: config/confighttp/confighttp.go [hunk 6]
```diff
@@ -286,6 +288,14 @@ type ServerConfig struct {
 	ResponseHeaders map[string]configopaque.String `mapstructure:"response_headers"`
 }
 
+// NewDefaultServerConfig creates new ServerConfig with default values set
+func NewDefaultServerConfig() ServerConfig {
+
```
Comment (TylerHelmuth) [inline config/confighttp/confighttp.go:293]:
```suggestion
```

File: config/confighttp/confighttp.go [hunk 7]
```diff
@@ -286,6 +288,14 @@ type ServerConfig struct {
...
+	return ServerConfig{
+		ResponseHeaders: make(map[string]configopaque.String),
```
Comment (TylerHelmuth) [inline config/confighttp/confighttp.go:295]: This should use configtls.NewDefaultServerConfig and configauth.NewDefaultAuthentication and NewDefaultCORSConfig

File: config/confighttp/confighttp.go [hunk 8]
```diff
@@ -107,10 +107,12 @@ type ClientConfig struct {
 // We encourage to use this function to create an object of ClientConfig.
 func NewDefaultClientConfig() ClientConfig {
 	// The default values are taken from the values of 'DefaultTransport' of 'http' package.
+
 	maxIdleConns := 100
 	idleConnTimeout := 90 * time.Second
 
 	return ClientConfig{
+		Headers:         make(map[string]configopaque.String),
```
Comment (TylerHelmuth) [inline config/confighttp/confighttp.go:115]: We should also use configtls.NewDefaultClientConfig and configauth.NewDefaultAuthentication
